shared/desktop/yarn-helper/action-creator-creator.js

```diff
@@ -0,0 +1,123 @@
+// @flow
+import prettier from 'prettier'
+import path from 'path'
+import json5 from 'json5'
+import fs from 'fs'
+
+type Payload = Object
+type ErrorPayload = {
+  canError: string,
+  ...Payload,
+}
+
+type ActionNS = string
+type ActionName = string
+type ActionDesc = Payload | ErrorPayload
+
+type Actions = {[key: ActionName]: ActionDesc}
+
+type FileDesc = {
+  prelude: string, // anything to prepend to our generated file
+  actions: Actions,
+}
+
+type CompileActionFn = (ns: ActionNS, actionName: ActionName, desc: ActionDesc) => string
+
+function compile(ns: ActionNS, {prelude, actions}: FileDesc): string {
+  return `// @flow
+/* eslint-disable */
+${prelude}
+
+type _ExtractReturn<B, F: (...args: any[]) => B> = B
+export type ReturnType<F> = _ExtractReturn<*, F>
+export type PayloadType<F> = $PropertyType<ReturnType<F>, 'payload'>
+
+// Constants
+${compileActions(ns, actions, compileReduxTypeConstant)}
+
+// Action Creators
+${compileActions(ns, actions, compileActionCreator)}
+
+// All Actions
+${compileAllActionsType(ns, actions)}
+  `
+}
+
+function compileAllActionsType(ns: ActionNS, actions: Actions): string {
+  return `export type Actions = ${Object.keys(actions)
+    .map(
+      (name: ActionName) =>
+        `ReturnType<typeof create${capitalize(name)}>` +
+        (actions[name].canError ? ` | ReturnType<typeof create${capitalize(name)}Error>` : '')
+    )
+    .join('|')}`
+}
+
+function compileActions(ns: ActionNS, actions: Actions, compileActionFn: CompileActionFn): string {
+  return Object.keys(actions)
+    .map((actionName: ActionName) => compileActionFn(ns, actionName, actions[actionName]))
+    .join('\n')
+}
+
+function capitalize(s: string): string {
+  return s[0].toUpperCase() + s.slice(1)
+}
+
+function actionReduxTypeName(ns: ActionNS, actionName: ActionName): string {
+  return `'${ns}:${actionName}'`
+}
+
+function printPayload(p: Object) {
+  return '{|' + Object.keys(p).map(key => `${key}: ${p[key]}`).join(',\n') + '|}'
+}
+
+function compileActionCreator(ns: ActionNS, actionName: ActionName, desc: ActionDesc) {
+  const {canError, ...noErrorPayload} = desc
+  return (
+    `export const create${capitalize(actionName)} = (payload: ${printPayload(noErrorPayload)}) => (
```

Comment:
might be worth having a talk sometime about what we want the name semantics of the action creators / types to be. cc @keybase/react-hackers